src/mono/mono/mini/mini-llvm.c

```diff
@@ -3752,6 +3766,27 @@ emit_entry_bb (EmitContext *ctx, LLVMBuilderRef builder)
 
 	ctx->alloca_builder = create_builder (ctx);
 
+#ifdef TARGET_WASM
+	/*
+	 * For GC stack scanning to work, allocate an area on the stack and store
+	 * every ref vreg into it after its written. Because the stack is scanned
+	 * conservatively, the objects will be pinned, so the vregs can directly
+	 * reference the objects, there is no need to load them from the stack
+	 * on every access.
+	 */
+	ctx->gc_var_indexes = g_new0 (int, cfg->next_vreg);
+	int ngc_vars = 0;
+	for (i = 0; i < cfg->next_vreg; ++i) {
+		if (vreg_is_ref (cfg, i)) {
+			ctx->gc_var_indexes [i] = ngc_vars + 1;
+			ngc_vars ++;
+		}
+	}
+
+	// FIXME: Count only live vregs
+	ctx->gc_pin_area = build_alloca_llvm_type_name (ctx, LLVMArrayType (IntPtrType (), ngc_vars), 0, "gc_pin");
```

Comment:
Is LLVM SROA smart enough to shrink this alloca if there are slots that are never written? 

Is LLVM dead code analysis smart enough to eliminate the stores since nothing ever reads them?

Will LLVM promote some of these stores to registers, defeating the what you're doing here?

I wonder if the gc pin area needs to be an array of volatile pointers.

Reply:
The stores to the area are still volatile so llvm is prevented from optimizing them.

Reply:
The stores in `emit_gc_pin` are volatile (the fourth argument to that function's call to `mono_llvm_build_store` is `TRUE`), so DCE shouldn't be a problem. Monotonically-ordered stores would be another option. Probably doesn't matter for wasm. C-style volatility isn't tracked in the LLVM IR type system.